src/chan_vese.jl

```diff
@@ -0,0 +1,252 @@
+    c₁, c₂ = calculate_averages(img, H𝚽)
+
+    if ndims(img) == 2
+        return @. -λ₁ * (img - c₁)^2 + λ₂ * (img - c₂)^2
+    elseif ndims(img) == 3
+        return -λ₁ .* sum((img .- c₁).^2, dims=3) .+ λ₂ .* sum((img .- c₂).^2, dims=3)
+    end
+end
+# H𝚽 = LazyArray(@~ @. 1. * (𝚽ⁿ > 0))
+function _calculate_averages(img::AbstractArray{T, N}, 𝚽ⁿ::AbstractArray{T, M}) where {T<:Real, N, M}
+    ∫H𝚽 = ∫H𝚽ⁱ = ∫u₀H𝚽 = ∫u₀H𝚽ⁱ = 0
+
+    for i in CartesianIndices(img)
+        H𝚽 = 1. * (𝚽ⁿ[i] > 0)
+        H𝚽ⁱ = 1. - H𝚽
+        ∫H𝚽 += H𝚽
+        ∫H𝚽ⁱ += H𝚽ⁱ
+        ∫u₀H𝚽 += img[i] * H𝚽
+        ∫u₀H𝚽ⁱ += img[i] * H𝚽ⁱ
+    end
+    if ∫H𝚽 != 0
+        c₁ = ∫u₀H𝚽 ./ ∫H𝚽
+    end
+    if ∫H𝚽ⁱ != 0
+        c₂ = ∫u₀H𝚽ⁱ ./ ∫H𝚽ⁱ
+    end
+
+    return c₁, c₂
+end
+
+function δₕ(x::AbstractArray{T,N}, h::Float64=1.0) where {T<:Real, N}
+    return @~ @. h / (h^2 + x^2)
+end
+
+function initial_level_set(shape::Tuple)
+    x₀ = reshape(collect(0:shape[begin]-1), shape[begin], 1)
+    y₀ = reshape(collect(0:shape[begin+1]-1), 1, shape[begin+1])
+    𝚽₀ = @. sin(pi / 5 * x₀) * sin(pi / 5 * y₀)
+end
+
+function chan_vese(img::GenericGrayImage;
+                    μ::Float64=0.25,
+                    λ₁::Float64=1.0,
+                    λ₂::Float64=1.0,
+                    tol::Float64=1e-3,
+                    max_iter::Int64=500,
+                    Δt::Float64=0.5,
+                    reinitial_flag::Bool=false) #where {T<:Real, N}
+    img = float64.(channelview(img))
+    iter = 0
+    h = 1.0
+    m, n = size(img)
+    s = m * n
+    𝚽ⁿ = initial_level_set((m, n)) # size: m * n
+    del = tol + 1
+    img .= img .- minimum(img)
+
+    if maximum(img) != 0
+        img .= img ./ maximum(img)
+    end
+
+    diff = 0
+    H𝚽 = similar(𝚽ⁿ)
+    u₀H𝚽 = similar(img)
+    ∫u₀ = sum(img)
+    𝚽ᵢ₊ᶜ = zeros(m, 1)
+
+
+
+    while (del > tol) & (iter < max_iter)
+        ϵ = 1e-16
+
+        @. H𝚽 = 1. * (𝚽ⁿ > 0) # size = (m, n)    
+        @. u₀H𝚽 = img * H𝚽 # size = (m, n) or (m, n, 3)
+
+        ∫H𝚽 = sum(H𝚽)
+        ∫u₀H𝚽 = sum(u₀H𝚽) # (1,)
+        ∫H𝚽ⁱ = s - ∫H𝚽
+        ∫u₀H𝚽ⁱ = ∫u₀ - ∫u₀H𝚽
+
+        if ∫H𝚽 != 0
+            c₁ = ∫u₀H𝚽 ./ ∫H𝚽
+        end
+        if ∫H𝚽ⁱ != 0
+            c₂ = ∫u₀H𝚽ⁱ ./ ∫H𝚽ⁱ
+        end
+
+        ind = CartesianIndices(reshape(collect(1 : 9), 3, 3)) .- CartesianIndex(2, 2)
+        𝚽ⱼ₊ = 0
+
+        for y in 1:n-1
+            𝚽ⱼ₊ = 0
+            for x in 1:m-1
+                i = CartesianIndex(x, y)
+                𝚽₀ = 𝚽ⁿ[i]
+                u₀ = img[i]
+                𝚽ᵢ₋ = 𝚽ᵢ₊ᶜ[i[1]]
+                𝚽ᵢ₊ᶜ[i[1]] = 𝚽ᵢ₊ = 𝚽ⁿ[i + ind[2, 3]] - 𝚽₀ # except i[2] = n
+                𝚽ⱼ₋ = 𝚽ⱼ₊
+                𝚽ⱼ₊ = 𝚽ⁿ[i + ind[3, 2]] - 𝚽₀ # except i[2] = m
+                𝚽ᵢ = 𝚽ᵢ₊ + 𝚽ᵢ₋
+                𝚽ⱼ = 𝚽ⱼ₊ + 𝚽ⱼ₋
+                t1 = 𝚽₀ + 𝚽ᵢ₊
+                t2 = 𝚽₀ - 𝚽ᵢ₋
+                t3 = 𝚽₀ + 𝚽ⱼ₊
+                t4 = 𝚽₀ - 𝚽ⱼ₋
+
+                C₁ = 1. / sqrt(ϵ + 𝚽ᵢ₊^2 + 𝚽ⱼ^2 / 4.)
```

Comment:
Not sure I follow what's happening here. Can you include a citation of the paper you're following in the docstring and then reference particular equations in the paper via comments?

Reply:
```julia
# Previous code
    𝚽⁺ = padarray(𝚽ⁿ, Pad(1, 1))
    r = axes(𝚽⁺)
    @views 𝚽ᵢ₊ = 𝚽⁺[first(r[1]) + 1:last(r[1]) - 1, first(r[2]) + 2:last(r[2])] .- 𝚽⁺[first(r[1]) + 1:last(r[1]) - 1, first(r[2]) + 1:last(r[2]) - 1]
    @views 𝚽ᵢ₋ = 𝚽⁺[first(r[1]) + 1:last(r[1]) - 1, first(r[2]) + 1:last(r[2]) - 1] .- 𝚽⁺[first(r[1]) + 1:last(r[1]) - 1, first(r[2]):last(r[2]) - 2]
    @views 𝚽ᵢ  = (𝚽⁺[first(r[1]) + 1:last(r[1]) - 1, first(r[2]) + 2:last(r[2])] .- 𝚽⁺[first(r[1]) + 1:last(r[1]) - 1, first(r[2]):last(r[2]) - 2]) / 2.

    @views 𝚽ⱼ₊ = 𝚽⁺[first(r[1]) + 2:last(r[1]), first(r[2]) + 1:last(r[2]) - 1] .- 𝚽⁺[first(r[1]) + 1:last(r[1]) - 1, first(r[2]) + 1:last(r[2]) - 1]
    @views 𝚽ⱼ₋ = 𝚽⁺[first(r[1]) + 1:last(r[1]) - 1, first(r[2]) + 1:last(r[2]) - 1] .- 𝚽⁺[first(r[1]):last(r[1]) - 2, first(r[2]) + 1:last(r[2]) - 1]
    @views 𝚽ⱼ  = (𝚽⁺[first(r[1]) + 2:last(r[1]), first(r[2]) + 1:last(r[2]) - 1] .- 𝚽⁺[first(r[1]):last(r[1]) - 2, first(r[2]) + 1:last(r[2]) - 1]) / 2.
```

```julia
# Current code
     𝚽ᵢ₊ᶜ = zeros(m, 1)
     𝚽ⱼ₊ = 0
     ...
     for y in 1:n-1
            𝚽ⱼ₊ = 0
            for x in 1:m-1
                i = CartesianIndex(x, y)
                𝚽₀ = 𝚽ⁿ[i]
                u₀ = img[i, :]
                𝚽ᵢ₋ = 𝚽ᵢ₊ᶜ[i[1]]
                𝚽ᵢ₊ᶜ[i[1]] = 𝚽ᵢ₊ = 𝚽ⁿ[i + ind[2, 3]] - 𝚽₀ # except i[2] = n
                𝚽ⱼ₋ = 𝚽ⱼ₊
                𝚽ⱼ₊ = 𝚽ⁿ[i + ind[3, 2]] - 𝚽₀ # except i[2] = m
                𝚽ᵢ = 𝚽ᵢ₊ + 𝚽ᵢ₋
                𝚽ⱼ = 𝚽ⱼ₊ + 𝚽ⱼ₋
                ...
            end
            i = CartesianIndex(m, y)
            𝚽₀ = 𝚽ⁿ[i]
            u₀ = img[i, :]
            𝚽ᵢ₋ = 𝚽ᵢ₊ᶜ[i[1]]
            𝚽ᵢ₊ᶜ[i[1]] = 𝚽ᵢ₊ = 𝚽ⁿ[i + ind[2, 3]] - 𝚽₀ # except i[2] = n
            𝚽ⱼ₋ = 𝚽ⱼ₊
            𝚽ⱼ₊ = 0 # except i[2] = m
            𝚽ᵢ = 𝚽ᵢ₊ + 𝚽ᵢ₋
            𝚽ⱼ = 𝚽ⱼ₊ + 𝚽ⱼ₋
            ...
     end
     ...
```
Some parts in this codes is dissimilar with the paper is because I try to avoid using `padarray` in the main cycle. Particularly, when we calculating forward diff `𝚽ᵢ₊, 𝚽ⱼ₊` and backward diff `𝚽ᵢ₋, 𝚽ⱼ₋`  we'll meet the boundary problems. In Python or Matlab's implementation of Chan Vese, `pad` is always used to get a bigger size image which helps us to solve boundary problems. However, I think this will cause lots of allocations, so I find some ways to avoid the boundary problems without using a bigger padded image.

Further more, take each column under consideration, the forward difference of i th column is equal to the backward difference of i + 1 th column. It is similar when we calculate the forward and backward difference of each row. So in current codes, I use `𝚽ⱼ₋ = 𝚽ⱼ₊, 𝚽ⱼ₊ = 𝚽ⁿ[i + ind[3, 2]] - 𝚽₀` to renew forward and backward difference for each pixel, so that we just need to calculate the forward difference, the value of backward difference comes from previous iteration. Since Julia's matrix is column major, we'll first loop along column direction, then along row direction. Therefore, when calculating the difference along row direction, the difference term of pixel (i, j) and (i, j + 1) are not calculated continuously, so I preallocate `𝚽ᵢ₊ᶜ = zeros(m, 1)` to save the values `𝚽ᵢ₊` which means forward difference of i th column. When the iteration comes to i + 1 th column, we just need to call `𝚽ᵢ₊ᶜ` to get the values of backward difference. Particularly, since backward diff for the first column is 0, so `𝚽ᵢ₊ᶜ` is initialized to be 0. The last row and last column is also with some difference when calculating forward diff, so I put them out of the `for` cycle.

After finding ways to optimize the calculation of forward diff and backward diff, some other problems can be solved simultaneously.
*Example:*
```julia
t1 = 𝚽₀ + 𝚽ᵢ₊
t2 = 𝚽₀ - 𝚽ᵢ₋
t3 = 𝚽₀ + 𝚽ⱼ₊
t4 = 𝚽₀ - 𝚽ⱼ₋
```
Take `t1` as an example, for matrix `𝚽ⁿ` whose size is `(256, 256)`, `t1` represents the term `𝚽ⁿ[i + 1, j]`. In condition that `i  = 256`, we meet the boundary problem because `𝚽ⁿ[257, j]` is unavailable. Moreover, it's inefficient to do lots of `if` judgement when calculating. Particularly, we can find that `𝚽ᵢ₊[i, j] = 𝚽ⁿ[i + 1, j] - 𝚽ⁿ[i, j]`, as a result, we can get the value of `𝚽ⁿ[i + 1, j]` by calculating `𝚽ᵢ₊[i, j] + 𝚽ⁿ[i, j]`. This is  the derivation of code `t1 = 𝚽₀ + 𝚽ᵢ₊`.

Reply:
These are the kind of comments that could be in the code itself. A reference to the particular Equation number that you are most similar to would ensure that readers know *what* you're calculating, and then comments can describe *how* you're calculating it. This section is sufficiently opaque that some pointers like these are necessary.

For example, if I read this correctly, are you effectively computing `|∇ϕ|`? I don't see any `abs`, and some of the key players in their regularizers (e.g., `arctan`) also seem missing. As a consequence, the role of your Cs, K, and delta are a bit of work for anyone except you to figure out.

This is also a *lot* of code for what looks like a fairly simple operation. Are you sure that caching the forward differences makes sense? Often a single subtraction is faster than looking something up in an array. If you've not benchmarked, don't assume that caching helps; often when the number of computations is small, it's better to just recalculate. If on the contrary you've benchmarked and seen a major performance advantage, then I'd be much more persuadable.

Julia lets you do multidmensional finite-differences pretty easily:
```julia
julia> N = 3   # spatial dimensions in the array (3 shows generality, but N = 2 is unchanged)
3

julia> ΔI = ntuple(d -> CartesianIndex(ntuple(i -> i == d ? 1 : 0, N)), N)
(CartesianIndex(1, 0, 0), CartesianIndex(0, 1, 0), CartesianIndex(0, 0, 1))

julia> # Forward |∇ϕ|
       ϕ0 = ϕ[i]; dϕ = sum(abs, map(Δi -> ϕ(i+Δi) - ϕ0, ΔI))
```
Short and readable, no?